lib/modules/featurePenaltyBox.js

```diff
@@ -0,0 +1,161 @@
+	},
+	suspendFeatures: {
+		type: 'boolean',
+		value: false,
+		description: 'resPenaltyBoxSuspendFeaturesDesc',
+	},
+	features: {
+		description: 'resPenaltyBoxFeaturesDesc',
+		type: 'table',
+		advanced: true,
+		addRowText: 'manually register feature',
+		fields: [
+			{
+				name: 'moduleID',
+				type: 'text',
+			},
+			{
+				name: 'featureID',
+				type: 'text',
+			},
+			{
+				name: 'monitoring',
+				type: 'boolean',
+				value: true,
+			},
+			{
+				name: 'penalty',
+				type: 'text',
+				value: '0',
+			},
+		],
+		value: [],
+	},
+};
+
+export function alterFeaturePenalty(moduleID: string, featureID: string, valueDelta: number) {
+	if (isNaN(parseInt(valueDelta, 10))) {
+		console.warn('Could not alter penalty for', moduleID, featureID, ' - bad value:', valueDelta);
+		return MIN_PENALTY;
+	}
+	if (!Modules.isEnabled(module.moduleID)) return MIN_PENALTY;
+	const value = getOrAddFeatures(moduleID, featureID);
+	if (!value.monitoring) return MIN_PENALTY;
+	value.penalty = Math.min(Math.max(value.penalty + valueDelta, MIN_PENALTY), MAX_PENALTY);
+	Options.set(module, 'features', module.options.features.value);
+	if (value.penalty >= MAX_PENALTY) {
+		suspendFeature(moduleID, featureID);
+	}
+	return value.penalty;
+}
+
+function stopMonitoringFeature(moduleID, featureID) {
+	const value = getOrAddFeatures(moduleID, featureID);
+	value.monitoring = false;
+	value.penalty = MIN_PENALTY;
+	Options.set(module, 'features', module.options.features.value);
+}
+
+export function getFeaturePenalty(moduleID: string, featureID: string) {
+	if (!Modules.isEnabled(module.moduleID)) return MIN_PENALTY;
+	const value = getOrAddFeatures(moduleID, featureID);
+	if (!value.monitoring) return MIN_PENALTY;
+	return value.penalty;
+}
+
+type DelayOption = {
+	value: number | string,
+	default?: number | string,
+	penalizedValue?: number,
+};
+
+export function penalizedDelay(moduleID: string, featureID: string, delay: DelayOption) {
+	if (!module.options.delayFeatures.value) {
+		return parseInt(delay.value, 10);
+	}
+	if (delay.value !== delay.default) {
+		return parseInt(delay.value, 10);
+	}
+	const penalty = getFeaturePenalty(moduleID, featureID);
+	if (!penalty || MIN_PENALTY >= penalty) {
+		return parseInt(delay.value, 10);
+	}
+
+	const max = parseInt(delay.penalizedValue || delay.value, 10) * 6;
+	const initial = parseInt(delay.default, 10);
+	const position = penalty / 100;
+
+	return Math.min(max, (max - initial) * position + initial);
+}
+
+function getOrAddFeatures(moduleID, featureID) {
+	const value = Options.table.getMatchingValueOrAdd(module, 'features', { moduleID, featureID });
+	const obj = mapScalarToObject(module.options.features, value);
+	obj.penalty = parseInt(obj.penalty, 10) || 0;
+	return obj;
+}
+
+async function suspendFeature(moduleID, featureID) {
+	if (!module.options.suspendFeatures.value) {
+		return;
+	}
+
+	const featureModule = Modules.get(moduleID);
+	const optionKey = featureID;
```

Comment:
I had originally had two motivations behind `featureID` abstraction:

1. Affects related options, e.g. `hoverDelay` for pop-up menus
1. Could be the whole module
2. Potentially support for other types of options besides booleans.

On reconsidering, 

1. There's typically a main option related options are always dependent on the primary option
2. It's easy to add a top-level option, or add `export const MODULE_FEATUREID = '---'` to support "module as feature"
3. For handling other types of options, we could extend the `penalizedDelay / option.penalizedValue` paradigm to other types, e.g. for enums `options.penalizedValue = [ 'disabled', 'partly-enabled-value', 'mostly-enabled-value', 'fully-enabled' ]` and map the penalty score onto that.

tldr: `s/featureID/optionKey/g`